src/Decompiler/Evaluation/ComplicatedNotRule.cs

```diff
@@ -0,0 +1,86 @@
+#region License
+/*
+ * Copyright (C) 2020 Sven Almgren.
+ *
+ * This program is free software; you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License as published by
+ * the Free Software Foundation; either version 2, or (at your option)
+ * any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program; see the file COPYING.  If not, write to
+ * the Free Software Foundation, 675 Mass Ave, Cambridge, MA 02139, USA.
+ */
+#endregion
+
+using Reko.Core;
+using Reko.Core.Expressions;
+using Reko.Core.Operators;
+using Reko.Core.Types;
+using System;
+
+namespace Reko.Evaluation
+{
+    /// <summary>
+    /// Rule that matches (0 - (-EXPRESSION == 0) + 1) and generates (!EXPRESSION).
+    /// This is a solution for a very specific sequence of instructions.
+    ///
+    /// Example x86 assembler:
+    /// neg ax
+    /// sbb ax, ax
+    /// inc ax
+    ///
+    /// Example result:
+    /// turn func(0x00 - (-MyVariable == 0x00) + 0x01) into func(!MyVariable)
+    /// </summary>
+    public class ComplicatedNotRule
+    {
+        private readonly EvaluationContext ctx;
+        private DataType? dataType;
+        private Expression? expression;
+
+        public ComplicatedNotRule(EvaluationContext ctx)
+        {
+            this.ctx = ctx;
+        }
+
+        public bool Match(BinaryExpression binExp)
+        {
+            if (binExp.Operator != Operator.IAdd)
+                return false;
+
+            if (!(binExp.Right is Constant rightConstant) || rightConstant.ToInt64() != 1)
+                return false;
+
+            if (!(binExp.Left is BinaryExpression leftExpression) || leftExpression.Operator != Operator.ISub)
+                return false;
+
+            if (!leftExpression.Left.IsZero)
+                return false;
+
+            if (!(leftExpression.Right is BinaryExpression middleExpression) || middleExpression.Operator != Operator.Eq)
+                return false;
+
+            if (!middleExpression.Right.IsZero)
+                return false;
+
+            if (!(middleExpression.Left is UnaryExpression unaryExpression) || unaryExpression.Operator != Operator.Neg)
```

Comment:
Consider supporting `0 - (exp == 0) + 1` as well as `0 - (-exp == 0) + 1`. That is, change this logic to:
```C#
if (middleExpression.Left is UnaryExpression unary && unaryExpression.Operator == Operator.Neg) 
{
    expression = unaryExpresion.Expression;
}
else 
{ 
    expression = middleExpression.Left;
}
```

Reply:
I'll look into it later tonight, good point! The reason I only supported -exp is that this is how they setup `sbb` to _do it's thing_ (the part about x86 specific). Do you have any samples that is affected by this that I can look at too, to make sure it's correct? Otherwise I'll just look into your example and make sure my case still works :P

Reply:
I'm seeing _similar_ code sequences in ~/subjects/Elf/x86-65/ls: for instance, the basic block `l000000000040A545` looks like this:
```
	and	edi,01
	mov	byte ptr [rsi+0A],20
	mov	byte ptr [rsi+0B],00
	cmp	edi,01
	sbb	eax,eax
	and	eax,E0
	add	eax,74
	mov	[rsi+09],al
	ret
```
Clearly a variant of the "subtract carry flag from register to yield either all 0's or all 1's" trick is being used here to generate either a lower case 't' (U+0074) or an upper case 'T' (U+0054). But this is not the same as your example.

In ~/subjects/MZ-x86/BENCHFN.EXE, I see in basic block `l0800_0A87` the following code:
```
	neg	cx
	sbb	ax,ax
	push	ax
```
which is _similar_ but again not the same as your sample.

~/subjects/PE/x86/ExeDll/Executable.exe has the end of basic block `l0040178F` looking like this:
```
; preceding instructions deleted
	neg	ebx
	lea	eax,[ebp-58]
	mov	[ebp-08],eax
	lea	eax,[ebp-00000324]
	sbb	bl,bl
	mov	[ebp-04],eax
	inc	bl
	call	dword ptr [00402038]
```
Notice the interleaving of instructions before the `call` instruction. Reko is currently (without your fix) generating this:
```
	byte bl_90 = 0<8> - (IsDebuggerPresent() == 1<32>)
```

Reply:
I removed that completely from this part and added two new rules, one simplifying unary expressions of the type (!-exp) into (!exp) and a binary that convert (-exp == 0) into (exp == 0). I don't believe I've seen a platform where the sign of an integer changes if it evaluates to zero or not.

The only thing that I could think of if the hardware in question would raise an exception if a value overflowed when changing the sign (like abs(MAX_INT) < abs(MIN_INT)). But I think other simplifications would ruin that too, so I don't think it's a loss.